### ISSUE TITLE
Incorrect suggestion in ptr_arg

### ISSUE THREAD

**Issue by cdisselkoen** [rust-lang/rust-clippy#4393]
For the following code, clippy's `ptr_arg` lint makes a suggestion which, when applied, results in a compiler error.

[Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=55048166d1a9b6ab4bff6d14051d873a)

Specifically, clippy suggests
```
warning: writing `&String` instead of `&str` involves a new object where a slice will do.
  --> src/lib.rs:61:32
   |
61 |     pub fn do_lookup(&self, k: &String) -> Option<usize> {
   |                                ^^^^^^^ help: change this to: `&str`
   |
   = note: `#[warn(clippy::ptr_arg)]` on by default
```
but changing the type of `k` to `&str` results in a compiler error.

The "Known Problems" section for `ptr_arg` [here](https://rust-lang.github.io/rust-clippy/master/index.html#ptr_arg) does include the caveat

> Other functions called from this function taking a `&String` or `&Vec` argument may also fail to compile if you change the argument. Applying this lint on them will fix the problem, but they may be in other crates.

but that is not quite what is happening here. Technically, `MyMap::get()` does have a `&String` in its signature (through the type parameter `K` which is instantiated as `String`), but `K` cannot be changed to `str` due to the `Sized` restriction.  So the problem cannot be fixed simply by applying the lint recursively.

I suppose the lint should probably be silenced in the case that there is a `Sized` restriction on the parameter type?

Clippy version: `clippy 0.0.212 (082cfa79 2019-06-07)`

**Comment by flip1995** [comment]
```rust
pub struct TwoMaps<'a> {
    map1: MyMap<&'a str, usize>,
    map2: MyMap<&'a str, usize>,
}

impl<'a> TwoMaps<'a> {
    pub fn do_lookup(&self, k: &str) -> Option<usize> {
        let v1 = self.map1.get(&k).copied()?;
        let v2 = self.map2.get(&k).copied()?;
        Some(v1 + v2)
    }
}
```

This would be a possible fix. If you want the key to be a `String`, this would be a possible fix:
```rust
impl TwoMaps {
    pub fn do_lookup(&self, k: &str) -> Option<usize> {
        let key = k.to_string()
        let v1 = self.map1.get(&key).copied()?;
        let v2 = self.map2.get(&key).copied()?;
        Some(v1 + v2)
    }
}
```

I do concede, that the second fix just shifts the problem. But shifting the building of a `String` to the very end of the chain may still be better style since you only have to build a `String` when it's *really* necessary.

**Comment by cdisselkoen** [comment]
Thanks for the comment.  You've convinced me that the lint is correct and it's not a bug.  I do prefer your first solution to the second - it may be that the caller already has a `String` object, in which case the second suggestion introduces a needless copy of the string if I'm not mistaken.  But either way, I see that the lint is still appropriate.  I'll close this.

Thanks for your help!

**Comment by flip1995** [comment]
Not sure about this, but if the caller already has a `String` object and passes it as a `&String`, I would expect that the compiler optimizes the `to_string` call away. But I don't know if this is a thing.
